### PR TITLE
test: fix test false positive, add types, use pathlib, change from format to fstrings

### PR DESCRIPTION
**Pull request by jackgerrits**
- Fix false positive test passing
  - Before this change `difflib` was used to determine if there were changes. If there was a diff, then we compared each different line to see if there were float differences that were close enough. However, if the diff differed because of ordering of lines this method would falsely accept the diff as close enough. This is resolved by not using `difflib` for the actual comparison and only using it to generate a diff should they be deemed to be different with a line by line comparison
- Add types
- Use `pathlib.Path` instead of string based paths whenever possible
- Use f-strings now that we can safely assume at least Python 3.6 is being used

### CODE REVIEW

File: test/run_tests.py [hunk 1]
```diff
@@ -42,20 +42,87 @@ class Result(Enum):
     SKIPPED = 3
 
 
-def try_decode(binary_object):
+class Completion:
```
Comment (ataymano) [inline test/run_tests.py:45]: Looks like type hinting is forgotten here

Comment (jackgerrits) [reply test/run_tests.py:45]: I explicitly typed `completed` in `__init__`, and left `lock` and `condition` to be implicitly typed based on their initialization

Comment (ataymano) [reply test/run_tests.py:45]: not sure if I understand. I was all methods signatures - id/success are unhinted.

File: test/run_tests.py [hunk 2]
```diff
@@ -42,20 +42,87 @@ class Result(Enum):
...
+        checks: Dict[str, Union[StatusCheck, DiffCheck]],
+        skip_reason: Optional[str] = None,
+    ):
+        self.id = id
+        self.result = result
+        self.checks = checks
+        self.skip_reason = skip_reason
+
+
+def try_decode(binary_object) -> str:
```
Comment (ataymano) [inline test/run_tests.py:112]: and the same here - type hint for binary_object?

File: test/run_tests.py (hunk 1 above)
Comment (jackgerrits) [reply test/run_tests.py:45]: Oh youre so right i missed those, thanks! I misunderstood originally

File: test/run_tests.py (hunk 2 above)
Comment (jackgerrits) [reply test/run_tests.py:112]: I'm not actually sure what the type of binary object is, I'll look into that one

File: test/run_tests.py [hunk 3]
```diff
@@ -130,21 +200,25 @@ def is_line_different(output_line, ref_line, epsilon):
...
             return (
                 True,
-                "Floats don't match {} {}".format((output_token), (ref_token)),
+                f"Floats don't match {output_token} {ref_token}",
                 found_close_floats,
             )
         else:
             if output_token != ref_token:
+                print(output_tokens)
+                print(ref_tokens)
```
Comment (bassmang) [inline test/run_tests.py:209]: Are these prints meant to be here?

File: test/run_tests.py [hunk 4]
```diff
@@ -370,131 +373,89 @@ def run_command_line_test(
...
             if output_file == "stdout":
                 output_content = stdout
             elif output_file == "stderr":
                 output_content = stderr
             else:
-                output_file_working_dir = os.path.join(working_dir, output_file)
-                if os.path.isfile(output_file_working_dir):
-                    output_content = open(
-                        output_file_working_dir, "r", encoding="utf-8"
+                output_file_working_dir = current_test_working_dir / output_file
```
Comment (bassmang) [inline test/run_tests.py:394]: Cool didnt know this was possible

File: test/run_tests.py (hunk 3 above)
Comment (jackgerrits) [reply test/run_tests.py:209]: No they arent, thanks for catching!

File: test/run_tests.py (hunk 4 above)
Comment (jackgerrits) [reply test/run_tests.py:394]: Yeah pathlib is super nice

**Review by bassmang — APPROVED** [review]

**Pull request merged**